File: ring-android/app/src/main/java/cx/ring/settings/pluginssettings/PluginPreferencesDataStore.java
package cx.ring.settings.pluginssettings;

import androidx.annotation.Nullable;
import androidx.preference.PreferenceDataStore;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;

import static cx.ring.plugins.PluginUtils.listStringToStringList;
import static cx.ring.plugins.PluginUtils.stringListToListString;

public class PluginPreferencesDataStore extends PreferenceDataStore {

    private PluginDetails mPluginDetails;
    private Map<String, String> mPreferenceTypes = new HashMap<>();
    private Map<String, String> preferencesValues;
    private ReadWriteLock lock = new ReentrantReadWriteLock();

    public PluginPreferencesDataStore(PluginDetails pluginDetails) {
        mPluginDetails = pluginDetails;
        preferencesValues = mPluginDetails.getPluginPreferencesValues();
    }

    public void addTomPreferenceTypes(Map<String, String> preferenceModel) {
        Lock writeLock = lock.writeLock();
        try {
            writeLock.lock();
            mPreferenceTypes.put(preferenceModel.get("key"), preferenceModel.get("type"));
        } finally {
            writeLock.unlock();
        }
    }

    @Override
    public void putBoolean(String key, boolean value) {
        Boolean boxedValue = value;
        boolean success = mPluginDetails.setPluginPreference(key, boxedValue.toString());
        if(success) {
            notifyPreferencesValuesChange();
        }
    }


    @Override
    public void putString(String key, @Nullable String value) {
        boolean success = mPluginDetails.setPluginPreference(key, value);
        if(success) {
            notifyPreferencesValuesChange();
        }
    }

    @Override
    public void putStringSet(String key, @Nullable Set<String> values) {
        if(values != null) {
            boolean success = mPluginDetails.setPluginPreference(key,
                    listStringToStringList(new ArrayList<>(values)));
            if(success) {
                notifyPreferencesValuesChange();
            }
        }
    }

    @Override
    public void putInt(String key, int value) {
        Integer boxedValue = value;
        boolean success = mPluginDetails.setPluginPreference(key, boxedValue.toString());
        if(success) {
            notifyPreferencesValuesChange();
        }
    }

    @Override
    public void putLong(String key, long value) {
        Long boxedValue = value;
        boolean success = mPluginDetails.setPluginPreference(key, boxedValue.toString());
        if(success) {
            notifyPreferencesValuesChange();
        }
    }

    @Override
    public void putFloat(String key, float value) {
        Float boxedValue = value;
        boolean success = mPluginDetails.setPluginPreference(key, boxedValue.toString());
        if(success) {
            notifyPreferencesValuesChange();
        }
    }

    @Nullable
    @Override
    public String getString(String key, @Nullable String defValue) {
        String returnValue = defValue;
        String value = getPreferencesValues().get(key);
        if (value != null) {
            returnValue = value;
        }
        return returnValue;
    }

    @Nullable
    @Override
    public Set<String> getStringSet(String key, @Nullable Set<String> defValues) {
        Set<String> returnValue =  defValues;

        String value = getPreferencesValues().get(key);

        if(value != null) {
            returnValue = new HashSet<>(stringListToListString(value));
        }

        return returnValue;
    }

    @Override
    public int getInt(String key, int defValue) {
        int returnValue = defValue;
        String value = getPreferencesValues().get(key);
        if (value != null) {
            returnValue = Integer.parseInt(value);
        }
        return returnValue;
    }

    @Override
    public long getLong(String key, long defValue) {
        long returnValue = defValue;
        String value = getPreferencesValues().get(key);
        if (value != null) {
            returnValue = Long.parseLong(value);
        }
        return returnValue;
    }

    @Override
    public float getFloat(String key, float defValue) {
        float returnValue = defValue;
        String value = getPreferencesValues().get(key);
        if (value != null) {
            returnValue = Float.parseFloat(value);
        }
        return returnValue;
    }

    @Override
    public boolean getBoolean(String key, boolean defValue) {
        boolean returnValue = defValue;
        String value = getPreferencesValues().get(key);
        if (value != null) {
            returnValue = value.equals("1");
        }
        return returnValue;
    }

    /**
     *  Updates the preferencesValues map
     *  Use locks since the PreferenceInteraction is asynchronous
     */
    public void notifyPreferencesValuesChange() {
        Lock writeLock = lock.writeLock();
        try {
            writeLock.lock();
            preferencesValues = mPluginDetails.getPluginPreferencesValues();
        } finally {
            writeLock.unlock();
        }
    }

    /**
     * Returns the preferencesValues
     * Use locks since the PreferenceInteraction is asynchronous
     * @return preferencesValues
     */
    private Map<String, String>  getPreferencesValues() {
        Lock readLock = lock.readLock();
        try {
            readLock.lock();
            return preferencesValues;
        } finally {
            readLock.unlock();
        }
    }
}

File: ring-android/app/src/main/java/cx/ring/settings/pluginssettings/PluginPreferencesDataStore.kt
package cx.ring.settings.pluginssettings

import androidx.preference.PreferenceDataStore
import java.util.concurrent.locks.ReadWriteLock
import java.util.concurrent.locks.ReentrantReadWriteLock

class PluginPreferencesDataStore(private val mPluginDetails: PluginDetails) : PreferenceDataStore() {
    private val mPreferenceTypes: MutableMap<String, String> = HashMap()

    private var preferencesValues: Map<String, String> = mPluginDetails.pluginPreferencesValues
    private val lock: ReadWriteLock = ReentrantReadWriteLock()

    fun addToPreferenceTypes(preferenceModel: Map<String, String>) {
        val writeLock = lock.writeLock()
        try {
            writeLock.lock()
            mPreferenceTypes[preferenceModel["key"]!!] = preferenceModel["type"]!!
        } finally {
            writeLock.unlock()
        }
    }

    override fun putBoolean(key: String, value: Boolean) {
        if (mPluginDetails.setPluginPreference(key, if (value) "1" else "0")) {
            notifyPreferencesValuesChange()
        }
    }

    override fun putString(key: String, value: String?) {
        if (mPluginDetails.setPluginPreference(key, value ?: "")) {
            notifyPreferencesValuesChange()
        }
    }

    override fun putStringSet(key: String, values: Set<String>?) {
        if (values != null) {
            if (mPluginDetails.setPluginPreference(key, values.joinToString(","))) {
                notifyPreferencesValuesChange()
            }
        }
    }

    override fun putInt(key: String, value: Int) {
        if (mPluginDetails.setPluginPreference(key, value.toString())) {
            notifyPreferencesValuesChange()
        }
    }

    override fun putLong(key: String, value: Long) {
        if (mPluginDetails.setPluginPreference(key, value.toString())) {
            notifyPreferencesValuesChange()
        }
    }

    override fun putFloat(key: String, value: Float) {
        if (mPluginDetails.setPluginPreference(key, value.toString())) {
            notifyPreferencesValuesChange()
        }
    }

    override fun getString(key: String, defValue: String?): String? {
        return getPreferencesValues()[key] ?: defValue
    }

    override fun getStringSet(key: String, defValues: Set<String>?): Set<String>? {
        return getPreferencesValues()[key]?.split(',')?.toSet() ?: defValues
    }

    override fun getInt(key: String, defValue: Int): Int {
        return getPreferencesValues()[key]?.toInt() ?: defValue
    }

    override fun getLong(key: String, defValue: Long): Long {
        return getPreferencesValues()[key]?.toLong() ?: defValue
    }

    override fun getFloat(key: String, defValue: Float): Float {
        return getPreferencesValues()[key]?.toFloat() ?: defValue
    }

    override fun getBoolean(key: String, defValue: Boolean): Boolean {
        var returnValue = defValue
        val value = getPreferencesValues()[key]
        if (value != null) {
            returnValue = value == "1"
        }
        return returnValue
    }

    /**
     * Updates the preferencesValues map
     * Use locks since the PreferenceInteraction is asynchronous
     */
    private fun notifyPreferencesValuesChange() {
        val writeLock = lock.writeLock()
        preferencesValues = try {
            writeLock.lock()
            mPluginDetails.pluginPreferencesValues
        } finally {
            writeLock.unlock()
        }
    }

    /**
     * Returns the preferencesValues
     * Use locks since the PreferenceInteraction is asynchronous
     * @return preferencesValues
     */
    private fun getPreferencesValues(): Map<String, String> {
        val readLock = lock.readLock()
        return try {
            readLock.lock()
            preferencesValues
        } finally {
            readLock.unlock()
        }
    }
}